protocol/transaction_validator.go

```diff
@@ -0,0 +1,317 @@
+		Value:           amount.String(),
+		Address:         to,
+		Contract:        actualContractAddress,
+		AccordingToSpec: false,
+		Valid:           true,
+	}, nil
+
+}
+
+func (t *TransactionValidator) validateToAddress(specifiedTo, actualTo string) bool {
+	if len(specifiedTo) != 0 && (strings.ToLower(specifiedTo) != strings.ToLower(actualTo) || !t.addresses[strings.ToLower(actualTo)]) {
+		return false
+	}
+
+	return t.addresses[actualTo]
+}
+
+func (t *TransactionValidator) validateEthereumTransfer(parameters *CommandParameters, transaction coretypes.Message) (*VerifyTransactionResponse, error) {
+	toAddress := strings.ToLower(transaction.To().Hex())
+
+	if !t.validateToAddress(parameters.Address, toAddress) {
+		return &VerifyTransactionResponse{}, nil
+	}
+	amount := transaction.Value()
+	if parameters.Value != "" {
+		advertisedAmount, ok := new(big.Int).SetString(parameters.Value, 10)
+		if !ok {
+			return nil, errors.New("can't parse amount")
+		}
+		return &VerifyTransactionResponse{
+			AccordingToSpec: amount.Cmp(advertisedAmount) == 0,
+			Valid:           true,
+			Value:           amount.String(),
+			Address:         toAddress,
+		}, nil
+
+	} else {
+		return &VerifyTransactionResponse{
+			AccordingToSpec: false,
+			Valid:           true,
+			Value:           amount.String(),
+			Address:         toAddress,
+		}, nil
+
+	}
+
+}
+
+type VerifyTransactionResponse struct {
+	Pending bool
+	// AccordingToSpec means that the transaction is valid,
+	// the user should be notified, but is not the same as
+	// what was requested, for example because the value is different
+	AccordingToSpec bool
+	// Valid means that the transaction is valid
+	Valid bool
+	// The actual value received
+	Value string
+	// The contract used in case of tokens
+	Contract string
+	// The address the transaction was actually sent
+	Address string
+
+	Message     *Message
+	Transaction *TransactionToValidate
+}
+
+// validateTransaction validates a transaction and returns a response.
+// If a negative response is returned, i.e `Valid` is false, it should
+// not be retried.
+// If an error is returned, validation can be retried.
+func (t *TransactionValidator) validateTransaction(message coretypes.Message, parameters *CommandParameters, from *ecdsa.PublicKey) (*VerifyTransactionResponse, error) {
+
+	fromAddress := types.BytesToAddress(message.From().Bytes())
+
+	err := t.verifyTransactionSignature(from, fromAddress, parameters.TransactionHash, parameters.Signature)
+	if err != nil {
+		t.logger.Error("failed validating signature", zap.Error(err))
+		return &VerifyTransactionResponse{}, nil
+	}
+
+	if len(message.Data()) != 0 {
+		t.logger.Debug("Validating token")
+		return t.validateTokenTransfer(parameters, message)
+	} else {
+		t.logger.Debug("Validating eth")
+		return t.validateEthereumTransfer(parameters, message)
+	}
+}
+
+func (t *TransactionValidator) ValidateTransactions(client EthClient) ([]*VerifyTransactionResponse, error) {
```

Comment:
What's the reason to pass `client` here instead of the constructor `NewTransactionValidator`?

Reply:
mainly because messenger takes care of initializing it (they share persistence), and does not have access to `EthClient` (which I understand we want to use interfaces in `protocol` and let `status-go` give a concrete implementation), do you think is ok?

Reply:
So, it's like this:
```
// in shhext service
client := ...
messenger.ValidateTransactions(client)

// in Messenger
func (m *Messenger) ValidateTransactions(client) {
    validator := NewTransactionValidator()
    responses, err := validator.ValidateTransactions(client)
}
```
So, it looks like it's totally fine to pass `client` to `NewTransactionValidator`.

But that's just a technicality which does not change anything except for making `TransactionValidator` methods a bit simpler.

> which I understand we want to use interfaces in protocol and let status-go give a concrete implementation

That's correct.

--- 

Alternative thing is that `NewMessanger` would take an option `WithEthClient` and the client would be stored in `Messenger`'s field. That would probably make the `Messenger` interface the most clear but not sure we can have `EthClient` instantiated where the `Messenger` is. 

Also, `EthClient` on `Messenger` could be optional and validating txs would always fail if the client is not provided. Would that be helpful if someone does not take care about txs? I am thinking here of bots or bridges which only care about public messages.

Reply:
going for the latter option